File: DSA/src/Questions/Array/ThreeSumClosest.java
package Questions.Array;

import java.util.Arrays;
/*
 https://leetcode.com/problems/3sum-closest/
 Given an integer array nums of length n and an integer target, find three integers in nums such that the sum is closest to target.

Return the sum of the three integers.

You may assume that each input would have exactly one solution.

 

Example 1:

Input: nums = [-1,2,1,-4], target = 1
Output: 2
Explanation: The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
Example 2:

Input: nums = [0,0,0], target = 1
Output: 0
Explanation: The sum that is closest to the target is 0. (0 + 0 + 0 = 0).
 

Constraints:

3 <= nums.length <= 500
-1000 <= nums[i] <= 1000
-104 <= target <= 104
 */
public class ThreeSumClosest {
public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int closestSum = nums[0] + nums[1] + nums[2]; 

        for (int i = 0; i < nums.length - 2; i++) {
            int start = i + 1;
            int end = nums.length - 1;

            while (start < end) {
                int sum = nums[i] + nums[start] + nums[end];

                if (Math.abs(target - sum) < Math.abs(target - closestSum)) {
                    closestSum = sum; 
                }

                if (sum < target) {
                    start++; 
                } else {
                    end--; // Decrement k to decrease the sum
                }
            }
        }

        return closestSum;
    }

    public static void main(String[] args) {
        ThreeSumClosest threeSumClosest = new ThreeSumClosest();
        int[] nums = {1, 2, 4, 8, 16, 32, 64, 128};
        int target = 82;
        System.out.println(threeSumClosest.threeSumClosest(nums, target));
    }
}

File: DSA/Python/Questions/Array/ThreeSumClosest.py
"""
# Question: Three Sum Closest
# Link: https://leetcode.com/problems/3sum-closest/

# Find three integers that sum closest to target value

# Time Complexity: O(n^2) using sorting and two pointers
# Space Complexity: O(1)

# Algorithm:
# 1. Sort array for efficient two-pointer approach
# 2. Fix first element and find pairs for remaining sum
# 3. Use two pointers to minimize difference with target
# 4. Track closest sum throughout iteration

# Key Components:
# - three_sum_closest(): Main function implementing closest sum finding
# - Two-pointer technique with difference tracking
# - Early termination on exact match
"""

from typing import List


class ThreeSumClosest:
    def three_sum_closest(self, nums: List[int], target: int) -> int:
        nums.sort()
        closest_sum = nums[0] + nums[1] + nums[2]

        for i in range(len(nums) - 2):
            start = i + 1
            end = len(nums) - 1

            while start < end:
                current_sum = nums[i] + nums[start] + nums[end]

                if abs(target - current_sum) < abs(target - closest_sum):
                    closest_sum = current_sum

                if current_sum < target:
                    start += 1
                else:
                    end -= 1

        return closest_sum


def main():
    solution = ThreeSumClosest()
    nums = [1, 2, 4, 8, 16, 32, 64, 128]
    target = 82
    result = solution.three_sum_closest(nums, target)
    print(result)


if __name__ == "__main__":
    main()
